### PR TITLE
Integrate DefaultAuthorizer with PolicyEvaluator

### PR DESCRIPTION
**Pull request by mithmatt**
### Describe your changes :

I worked on this because #1940 

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have performed a self-review of my own. 
- [x] I have tagged my reviewers below.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

Backend: @sureshms @harshach

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java [hunk 1]
```diff
@@ -66,19 +70,25 @@
...
+          @Override
+          public void logAfterExecution(StatementContext context) {
+            LOG.debug(
+                "sql {}, parameters {}, timeTaken {} ms",
+                context.getRenderedSql(),
+                context.getBinding().toString(),
+                context.getElapsedTime(ChronoUnit.MILLIS));
+          }
+        };
+    if (LOG.isDebugEnabled()) {
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java:89]: @sureshms Is it okay to enable SQL Logger if log level is set to debug?

Was contemplating removing commented out code, and figured we may need it for debugging

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/Entity.java [hunk 2]
```diff
@@ -138,12 +143,45 @@ public static EntityReference withHref(UriInfo uriInfo, EntityReference ref) {
       return null;
     }
     String entityName = getEntityNameFromObject(entity);
+    EntityRepository<T> entityRepository = getEntityRepository(entityName);
+    return entityRepository.getEntityInterface(entity);
+  }
+
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/Entity.java:149]: Adding a couple of new util functions.

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/policyevaluator/AttributeBasedFacts.java [hunk 3]
```diff
@@ -46,35 +45,29 @@ public boolean hasPermission() {
...
     return entityTags.stream().map(TagLabel::getTagFQN).collect(Collectors.toList());
   }
 
-  private static String getEntityType(Object object) {
-    // TODO: Fix the entityType fetch such that it is fetched from Repository or using a Util.
-    // This is done here now to facilitate prototyping and unit testing.
-    return object.getClass().getSimpleName().toLowerCase(Locale.ROOT);
-  }
-
-  private boolean validate() {
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/policyevaluator/AttributeBasedFacts.java:72]: Remove validate and use lombok NonNull Builder

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/util/EntityUtil.java [hunk 4]
```diff
@@ -391,6 +392,7 @@ public static boolean addFollower(
     return followers;
   }
 
+  @RequiredArgsConstructor
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/util/EntityUtil.java:395]: This enables

```
List<String> fieldsList = new ArrayList<>();
fieldsList.add("policyType");
Fields fields = new Fields(fieldsList);
```

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java [hunk 5]
```diff
@@ -123,6 +128,20 @@ public boolean hasPermissions(AuthenticationContext ctx, EntityReference owner)
     }
   }
 
+  @Override
+  public boolean hasPermissions(
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java:132]: TODO: Need to wire this up to some of the Resource APIs and write tests around authorization of Metadata Operation changes

**Comment by harshach** [comment]
@mithmatt this looks great. I think hasPermissions should have role check as well to see if the user is admin or bot and grant permissions to short circuit. All resources can call this method or we can build a decorator for it

**Comment by mithmatt** [comment]
> @mithmatt this looks great. I think hasPermissions should have role check as well to see if the user is admin or bot and grant permissions to short circuit. All resources can call this method or we can build a decorator for it

Agreed - making changes

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/Authorizer.java [hunk 6]
```diff
@@ -13,20 +13,28 @@
...
    * Check if the authenticated user has given permission on the target entity identified by the given resourceType and
    * resourceName
    */
   boolean hasPermissions(AuthenticationContext ctx, EntityReference entityReference);
 
+  /**
+   * Check if the authenticated user (subject) has permission to perform the {@link MetadataOperation} on the target
+   * entity (object).
+   */
+  boolean hasPermissions(AuthenticationContext ctx, EntityReference entityReference, MetadataOperation operation);
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/Authorizer.java:36]: The new API

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java (hunk 5 above)
Comment (mithmatt) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java:132]: Do this in next PR

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/SecurityUtil.java [hunk 7]
```diff
@@ -55,6 +56,20 @@ public static void checkAdminRoleOrPermissions(
     }
   }
 
+  public static void checkAdminRoleOrPermissions(
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/SecurityUtil.java:59]: @harshach 

This is the overloaded function that does checkAdmin, role and metadata operation. 
Going to attempt hooking this up to the resource APIs.

**Review by harshach — APPROVED** [review]

**Pull request merged**